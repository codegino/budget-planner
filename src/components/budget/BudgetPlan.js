import React, { PureComponent } from 'react';
import { Text, View, TouchableNativeFeedback, StyleSheet, FlatList, Platform, TouchableOpacity } from 'react-native';
import { connect } from 'react-redux';
import { Navigation } from 'react-native-navigation';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';
import _ from 'lodash';

import ItemSubMenu from '../../components/menu/ItemSubMenu';
import ProgressBar from './ProgressBar';
import colors from '../../constants/colors';
import { updateBudgetPlan, deleteBudgetPlan } from '../../store/actions/budgetPlan';


const styles = StyleSheet.create({
  container: {
    borderWidth: 2,
    borderRadius: 10,
    borderColor: colors.primaryDark,
    backgroundColor: colors.primary3,
    marginTop: 3,
    marginBottom: 3,
    marginHorizontal: 5,
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  budgetPlan: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  budgetPlanWrapper: {
    width: '100%',
    paddingLeft: 10,
    paddingRight: 20,
    justifyContent: 'space-between',
  },
  label: {
    width: '100%',
    textAlign: 'center',
    fontSize: 20,
    fontWeight: 'bold',
    color: 'white',
  },
  bottom: {
    marginTop: 3,
    marginBottom: 3,
  },
  bottomText: {
    textDecorationLine: 'underline',
    alignSelf: 'center',
    color: 'white',
  },
  itemDetail: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    borderRadius: 5,
    borderWidth: 1,
    padding: 5,
    marginTop: 5,
    borderColor: colors.primaryLight,
    backgroundColor: colors.primaryDark,
  },
  detailText: {
    fontSize: 15,
    color: 'white',
  },
});

type BudgetProps = {
  budgetPlan: {
    category: string,
    budget: number,
    expenses: number,
    date: string,
  },
  deleteBudgetPlan: Function,
  budgetPlanItems: Function,
}

class Budget extends PureComponent<BudgetProps> {
  state = {
    isMenuVisible: false,
    isExpanded: false,
  }

  onLongPressHandler = () => {
    this.setState(prevState => ({
      ...this.state,
      isMenuVisible: !prevState.isMenuVisible,
    }));
  }

  onEditItemHandler = () => {
    Navigation.showModal({
      screen: 'expenses.AddUpdateBudgetPlanModal',
      title: 'Update budget plan',
      animationType: 'slide-up',
      passProps: { budgetPlan: this.props.budgetPlan, mode: 'edit' },
    });

    this.setState(prevState => ({
      ...prevState,
      isMenuVisible: false,
    }));
  }

  onDeleteItemHandler = () => {
    const { budgetPlan } = this.props;
    this.props.deleteBudgetPlan(budgetPlan);

    this.setState(prevState => ({
      ...prevState,
      isMenuVisible: false,
    }));
  }

  onCloseHandler = () => {
    this.setState(() => ({
      ...this.state,
      isMenuVisible: false,
    }));
  }

  onToggleExpandHandler = () => {
    this.setState(prevState => ({
      ...prevState,
      isExpanded: !prevState.isExpanded,
    }));
  };


  render() {
    let menu = null;
    let details = null;

    const bottomText = this.state.isExpanded ? 'Show less' : 'Show more';

    const itemDetail = ({ item }) => (
      <View style={styles.itemDetail}>
        <Text style={[styles.detailText, { fontWeight: 'bold', width: '50%' }]}>{item.name}</Text>
        <Text style={[styles.detailText, { width: '20%' }]}>{item.price}</Text>
        <Text style={[styles.detailText, { width: '30%' }]}>{item.date}</Text>
      </View>
    );

    if (this.state.isExpanded) {
      details = (<FlatList
        data={this.props.budgetPlanItems(
          this.props.budgetPlan.date,
          this.props.budgetPlan.category,
        )}
        renderItem={itemDetail}
        keyExtractor={item => item.id}
      />);
    }

    if (this.state.isMenuVisible) {
      menu = (
        <ItemSubMenu
          onEdit={this.onEditItemHandler}
          onDelete={this.onDeleteItemHandler}
          onClose={this.onCloseHandler}
        />
      );
    }

    let color = colors.success;
    let emoticon = <Icon name="emoticon-excited" size={50} color={color} />;

    const progress = (this.props.budgetPlan.expenses / this.props.budgetPlan.budget) * 100;

    if (progress >= 90) {
      color = colors.danger;
      emoticon = <Icon name="emoticon-dead" size={50} color={color} />;
    } else if (progress >= 70) {
      color = colors.warning;
      emoticon = <Icon name="emoticon-sad" size={50} color={color} />;
    } else if (progress >= 50) {
      color = colors.notify;
      emoticon = <Icon name="emoticon-neutral" size={50} color={color} />;
    }

    const content = (
      <View style={styles.container}>
        <View style={styles.budgetPlanWrapper}>
          <View style={styles.budgetPlan}>
            <View style={{ width: '90%', justifyContent: 'space-between' }}>
              <Text style={styles.label}>{this.props.budgetPlan.category}</Text>
              <ProgressBar
                numerator={this.props.budgetPlan.expenses}
                denominator={this.props.budgetPlan.budget}
                proress={progress}
                color={color}
              />
            </View>
            <View>
              {emoticon}
            </View>
          </View>
          {details}
          <TouchableNativeFeedback onPress={this.onToggleExpandHandler}>
            <View style={styles.bottom}>
              <Text style={styles.bottomText}>{bottomText}</Text>
            </View>
          </TouchableNativeFeedback>
          {menu}
        </View>
      </View>
    );

    if (Platform.OS === 'android') {
      return (
        <TouchableNativeFeedback onLongPress={this.onLongPressHandler} >
          {content}
        </TouchableNativeFeedback>
      );
    }
    return <TouchableOpacity onLongPress={this.onLongPressHandler}>{content}</TouchableOpacity>;
  }
}

const mapStateToProps = state => ({
  budgetPlanItems: (month, category) => _.sortBy(_.filter(state.expenses.expenses, o => (
    month === o.date.substring(3, 11) && category === o.category
  )), 'date'),
});

const mapDispatchToProps = dispatch => ({
  deleteBudgetPlan: item => dispatch(deleteBudgetPlan(item)),
  updateBudgetPlan: budgetPlan => dispatch(updateBudgetPlan(budgetPlan)),
});


export default connect(mapStateToProps, mapDispatchToProps)(Budget);
