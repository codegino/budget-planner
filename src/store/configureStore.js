import createSagaMiddleware from 'redux-saga';
import { createStore, combineReducers, applyMiddleware, compose } from 'redux';

import { watchAuth, watchExpenses, watchBudgetPlan } from './sagas/index';
import authReducer from './reducers/auth';
import uiReducer from './reducers/ui';
import expensesReducer from './reducers/expenses';
import budgetPlanReducer from './reducers/budgetPlan';

const sagaMiddleware = createSagaMiddleware();

const rootReducer = combineReducers({
  auth: authReducer,
  ui: uiReducer,
  expenses: expensesReducer,
  budgetPlan: budgetPlanReducer,
});

let composeEnhancers = compose;

if (__DEV__) {
  // eslint-disable-next-line
  composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
}

const configureStore = createStore(rootReducer, composeEnhancers(applyMiddleware(sagaMiddleware)));

export default () => configureStore;

sagaMiddleware.run(watchAuth);
sagaMiddleware.run(watchExpenses);
sagaMiddleware.run(watchBudgetPlan);
