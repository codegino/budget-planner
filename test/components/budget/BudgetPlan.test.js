import React from 'react';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import saga from 'redux-saga';
import { applyMiddleware } from 'redux';

import BudgetPlan from 'containers/budget/BudgetPlan';

const middlewares = [saga];

const state = {
  expenses: {
    expenses: [],
  },
};

it('will render', () => {
  const initialState = state;
  const mockStore = configureStore(initialState);

  const store = mockStore(initialState, applyMiddleware(middlewares));

  const budget = {
    category: 'test',
    budget: 100,
    expenses: 10,
    date: 'Jun-2010',
  };

  const budgetPlan = renderer.create(<BudgetPlan store={store} budgetPlan={budget} />);

  expect(budgetPlan.toJSON()).toMatchSnapshot();
});
