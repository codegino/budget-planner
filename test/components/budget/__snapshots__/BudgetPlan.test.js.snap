// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`will render 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  collapsable={undefined}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Object {
        "backgroundColor": "#1e583f",
        "borderColor": "#193e2c",
        "borderRadius": 10,
        "borderWidth": 2,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 3,
        "marginHorizontal": 5,
        "marginTop": 3,
      }
    }
  >
    <View
      style={
        Object {
          "justifyContent": "space-between",
          "paddingLeft": 10,
          "paddingRight": 20,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Object {
              "justifyContent": "space-between",
              "width": "90%",
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "white",
                "fontSize": 20,
                "fontWeight": "bold",
                "textAlign": "center",
                "width": "100%",
              }
            }
          >
            test
          </Text>
          <View
            style={
              Object {
                "backgroundColor": "#193e2c",
                "borderColor": "#5fbd98",
                "borderRadius": 5,
                "borderWidth": 2,
                "justifyContent": "center",
                "overflow": "hidden",
                "padding": 1,
                "width": "100%",
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "borderRadius": 3,
                    "flexDirection": "row",
                    "minHeight": 20,
                  },
                  Object {
                    "backgroundColor": "green",
                    "width": "10%",
                  },
                ]
              }
            />
            <View
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "position": "absolute",
                  "width": "100%",
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "alignSelf": "center",
                      "fontWeight": "bold",
                      "paddingHorizontal": 5,
                    },
                    Object {
                      "color": "white",
                    },
                  ]
                }
              >
                10
                /
                100
              </Text>
            </View>
          </View>
        </View>
        <View>
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "green",
                  "fontSize": 50,
                },
                undefined,
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
              ]
            }
          >
            ïš›
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#ffbcbc",
              "borderColor": "red",
              "borderWidth": 1,
              "height": 100,
              "justifyContent": "center",
              "margin": 10,
              "width": 300,
            },
            undefined,
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#333333",
              "margin": 20,
            }
          }
        >
          TouchableNativeFeedback is not supported on this platform!
        </Text>
      </View>
    </View>
  </View>
</View>
`;
